source/server/admin/admin_filter.cc

```diff
@@ -62,20 +63,37 @@ const Http::RequestHeaderMap& AdminFilter::getRequestHeaders() const {
 }
 
 void AdminFilter::onComplete() {
+  decoder_callbacks_->addDownstreamWatermarkCallbacks(watermark_callbacks_);
+
   const absl::string_view path = request_headers_->getPathValue();
   ENVOY_STREAM_LOG(debug, "request complete: path: {}", *decoder_callbacks_, path);
 
-  Buffer::OwnedImpl response;
   auto header_map = Http::ResponseHeaderMapImpl::create();
   RELEASE_ASSERT(request_headers_, "");
-  Http::Code code = admin_server_callback_func_(path, *header_map, response, *this);
+  handler_ = admin_handler_fn_(path, *this);
+  Http::Code code = handler_->start(*header_map);
   Utility::populateFallbackResponseHeaders(code, *header_map);
-  decoder_callbacks_->encodeHeaders(std::move(header_map),
-                                    end_stream_on_complete_ && response.length() == 0,
+  decoder_callbacks_->encodeHeaders(std::move(header_map), false,
                                     StreamInfo::ResponseCodeDetails::get().AdminFilterResponse);
 
-  if (response.length() > 0) {
-    decoder_callbacks_->encodeData(response, end_stream_on_complete_);
+  nextChunk();
+}
+
+void AdminFilter::nextChunk() {
+  if (!can_write_ || handler_ == nullptr) {
+    return;
+  }
+
+  Buffer::OwnedImpl response;
+  bool more_data = handler_->nextChunk(response);
+  bool end_stream = end_stream_on_complete_ && !more_data;
+  if (response.length() > 0 || end_stream) {
+    decoder_callbacks_->encodeData(response, end_stream);
+  }
+  if (more_data) {
+    decoder_callbacks_->dispatcher().post([this]() { nextChunk(); });
```

Comment:
All good questions and I'm evolving my understanding of how this ought to work, with much help from @alyssawilk .

One concern here is we don't want to starve other concurrent requests sharing the main thread, so we post as a way of doing cooperative multi-tasking (see https://github.com/envoyproxy/envoy/issues/16425). In H2 this would also allow other streams on the same connection to make progress, but here we are also thread-bound.

Another consideration is that admin does not make any upstream requests: all the data it needs is basically ready to go, but it may be too much for the downstream to swallow at once, and we don't want to buffer it all either.

I am not that concerned about spinning with a nextChunk not providing any data, but returning true, in the context of https://github.com/envoyproxy/envoy/pull/19693 -- that will not provide any empty chunks. Maybe I should disallow that case as I don't need to have it for that application.

In one intermediate version of this PR I didn't have nextChunk return a bool; it indicated completion by adding nothing to the response. That worked, but the call-sites turned out to be a lot messier.